libclc/amdgcn-amdhsa/libspirv/synchronization/barrier.cl

```diff
@@ -8,44 +8,47 @@
 
 #include <clc/clc.h>
 #include <spirv/spirv.h>
+#include <spirv/spirv_types.h>
 
-void __clc_amdgcn_s_waitcnt(unsigned flags);
+#define BUILTIN_FENCE(semantics, scope_memory)                                 \
+  if (semantics & 0x2)                                                         \
+    return __builtin_amdgcn_fence(__ATOMIC_ACQUIRE, scope_memory);             \
+  else if (semantics & 0x4)                                                    \
+    return __builtin_amdgcn_fence(__ATOMIC_RELEASE, scope_memory);             \
+  else if (semantics & 0x8)                                                    \
+    return __builtin_amdgcn_fence(__ATOMIC_ACQ_REL, scope_memory);             \
+  else if (semantics & 0x10)                                                   \
+    return __builtin_amdgcn_fence(__ATOMIC_SEQ_CST, scope_memory);             \
+  else                                                                         \
+    return __builtin_amdgcn_fence(__ATOMIC_ACQ_REL, scope_memory);
 
-// s_waitcnt takes 16bit argument with a combined number of maximum allowed
-// pending operations:
-// [12:8] LGKM -- LDS, GDS, Konstant (SMRD), Messages
-// [7] -- undefined
-// [6:4] -- exports, GDS, and mem write
-// [3:0] -- vector memory operations
-
-// Newer clang supports __builtin_amdgcn_s_waitcnt
-#if __clang_major__ >= 5
-#define __waitcnt(x) __builtin_amdgcn_s_waitcnt(x)
-#else
-#define __waitcnt(x) __clc_amdgcn_s_waitcnt(x)
-_CLC_DEF void __clc_amdgcn_s_waitcnt(unsigned) __asm("llvm.amdgcn.s.waitcnt");
-#endif
-
-_CLC_DEF _CLC_OVERLOAD void __mem_fence(cl_mem_fence_flags flags) {
-  if (flags & CLK_GLOBAL_MEM_FENCE) {
-    // scalar loads are counted with LGKM but we don't know whether
-    // the compiler turned any loads to scalar
-    __waitcnt(0);
-  } else if (flags & CLK_LOCAL_MEM_FENCE)
-    __waitcnt(0xff); // LGKM is [12:8]
+_CLC_DEF _CLC_OVERLOAD void __mem_fence(unsigned int scope_memory,
+                                        unsigned int semantics) {
+  switch ((enum Scope)scope_memory) {
+  default:
```

Comment:
```suggestion
  case CrossDevice:
```

That's the last missing enum value, no need for default, is the empty string actually doing the cross device mem-fence?

Reply:
Yeap, it's analogous to CUDA's `__nvvm_membar_sys` and used for an implementation of omp sytem fence: https://github.com/intel/llvm/blob/46046bd4312c3f81c300657097c480e3f56901df/openmp/libomptarget/DeviceRTL/src/Synchronization.cpp#L245